### PR TITLE
feat(gossip): compaction events crate

### PR DESCRIPTION
**Pull request by domodwyer**
Adds the "gossip_compaction" crate that layers compaction-specific logic over the underlying gossip crate to present a simpler API to callers.

Defines the new `CompactionEvents` gossip topic to separate out the compaction events from other messages.

---

* refactor(gossip): add Topic::CompactionEvents (044d5bfdc)
      
      This adds a separate compaction event gossip topic.

* feat(gossip): add gossip_compaction crate (c0b4a1087)
      
      Adds a crate that layers compaction-specific gossip types and
      abstractions over the underlying gossip transport for a nicer (and
      decoupled!) internal API.

### CODE REVIEW

File: gossip_compaction/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+//! round in its entirety, at the cost of discarding messages that exceed the
+//! maximum message size.
+//!
+//! Because the underlying gossip protocol does not provide guaranteed delivery,
+//! consumers of these messages SHOULD NOT expect to receive notifications for
+//! all events, so discarding large messages does not prevent a correctness
+//! problem - it does however reduce the effectiveness of the optimisations
+//! these notifications enable, and the frequency of message discards SHOULD be
+//! tracked to understand the scope of the problem.
+//!
```
Comment (domodwyer) [inline gossip_compaction/src/lib.rs:76]: This describes how callers should use this crate, the limitations the underlying transport imposes, and atomicity guarantees.

File: gossip_compaction/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        let got = node_b
+            .rx
+            .recv()
+            .with_timeout_panic(Duration::from_secs(5))
+            .await
+            .unwrap();
+
+        // Ensuring the content is identical
+        assert_eq!(got, want);
+    }
```
Comment (domodwyer) [inline gossip_compaction/src/lib.rs:259]: This integration test drives the whole thing over a real gossip link.

It also shows the API of this crate: `broadcast(event)` and receive event - that's it :+1:

File: gossip_compaction/src/rx.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+/// This type also provides a buffer between incoming events, and processing,
+/// preventing processing time from blocking the gossip reactor. Once the buffer
+/// is full, incoming events are dropped until space is made through processing
+/// of outstanding events. Dropping the [`CompactionEventRx`] stops the background
+/// event loop.
+#[derive(Debug)]
+pub struct CompactionEventRx {
+    tx: mpsc::Sender<Bytes>,
+    task: JoinHandle<()>,
+}
```
Comment (domodwyer) [inline gossip_compaction/src/rx.rs:49]: The `CompactionEventRx` type plugs into the underlying gossip `Dispatcher`, and is responsible for accepting incoming events and placing them into a processing queue. This decouples the latency of processing the event, from the gossip reactor.

A background task asynchronously pulls events from the processing queue, and delegates processing to the `ComapctionEventHandler`.

File: generated_types/src/lib.rs [hunk 4]
```diff
@@ -110,6 +110,9 @@ pub mod influxdata {
                 /// Parquet file creation notifications.
                 NewParquetFiles = 2,
 
+                /// Compaction round completion notifications.
+                CompactionEvents = 3,
```
Comment (savage-engineer) [inline generated_types/src/lib.rs:114]: Aha, the mystery number 3 has arrived! 🎩

File: gossip_compaction/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+//!             │ GossipHandle │          │    Dispatcher    │    │
+//!        │    └──────────────┘          └──────────────────┘
+//!                                                               │
+//!        └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+//! ```
+//!
+//! # Best Effort
+//!
+//! This underlying gossip subsystem is designed to provide best effort delivery
+//! of messages, and therefore best-effort delivery of parquet creation events,
```
Comment (savage-engineer) [inline gossip_compaction/src/lib.rs:53]:
```suggestion
//! of messages, and therefore best-effort delivery of compaction events,
```

Technically correct as is, but as there's the new `ParquetFileEvent` might as well disambiguate 🤷

File: gossip_compaction/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+        (a, b)
+    }
+
+    /// Ensure a ParquetFile can be round-tripped through the gossip layer.
+    ///
+    /// This acts as an integration test, covering the serialisation of parquet
+    /// file messages, passing into the gossip layer, topic assignment &
+    /// decoding, deserialisation of the parquet file message, and handling by
+    /// the new file event delegate abstraction defined by this crate.
```
Comment (savage-engineer) [inline gossip_compaction/src/lib.rs:208]:
```suggestion
    /// Ensure a CompactionEvent can be round-tripped through the gossip layer.
    ///
    /// This acts as an integration test, covering the serialisation of parquet
    /// file messages within a CompactionEvent message, passing into the gossip 
    /// layer, topic assignment & decoding, deserialisation of the compaction 
    /// event message, and handling by the new compaction event delegate 
    /// abstraction defined by this crate.
```

File: gossip_compaction/src/rx.rs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+{
+    async fn handle(&self, event: CompactionEvent) {
+        T::handle(self, event).await
+    }
+}
+
+/// An async gossip message dispatcher.
+///
+/// This type is responsible for deserialising incoming gossip
+/// [`Topic::NewParquetFiles`] payloads and passing them off to the provided
```
Comment (savage-engineer) [inline gossip_compaction/src/rx.rs:34]:
```suggestion
/// [`Topic::CompactionEvents`] payloads and passing them off to the provided
```

**Review by savage-engineer — APPROVED** [review]
Top stuff 👍 

Couple of doc suggestions

**Pull request merged**